pkg/sql/sqlinstance/instancestorage/instancestorage.go

```diff
@@ -239,3 +385,94 @@ func (s *Storage) ReleaseInstanceID(ctx context.Context, id base.SQLInstanceID)
 	}
 	return nil
 }
+
+// RunInstanceIDReclaimLoop runs a background task that allocates available
+// instance IDs and reclaim expired ones within the sql_instances table.
+func (s *Storage) RunInstanceIDReclaimLoop(
+	ctx context.Context,
+	stopper *stop.Stopper,
+	ts timeutil.TimeSource,
+	sessionExpirationFn func() hlc.Timestamp,
+) error {
+	return stopper.RunAsyncTask(ctx, "instance-id-reclaim-loop", func(ctx context.Context) {
+		ctx, cancel := stopper.WithCancelOnQuiesce(ctx)
+		defer cancel()
+
+		jitter := jitteredInterval
+		if s.TestingKnobs.JitteredIntervalFn != nil {
+			jitter = s.TestingKnobs.JitteredIntervalFn
+		}
+
+		timer := ts.NewTimer()
+		defer timer.Stop()
+		for {
+			timer.Reset(jitter(ReclaimLoopInterval.Get(&s.settings.SV)))
+			select {
+			case <-ctx.Done():
+				return
+			case <-timer.Ch():
+				timer.MarkRead()
+				if err := s.generateAvailableInstanceRows(ctx, sessionExpirationFn()); err != nil {
+					log.Warningf(ctx, "failed to generate available instance rows: %v", err)
+				}
+			}
+		}
+	})
+}
+
+// generateAvailableInstanceRows allocates available instance IDs, and store
+// them in the sql_instances table. When instance IDs are pre-allocated, all
+// other fields in that row will be NULL.
+//
+// TODO(jaylim-crl): Handle multiple regions in this logic. encodeRow has to be
+// updated with crdb_region. When we handle multiple regions, we have to figure
+// out where to get the list of regions, and ensure that we don't do a global
+// read for each region assuming that the number of pre-allocated entries is
+// insufficient. One global KV read and write would be sufficient for **all**
+// regions.
+func (s *Storage) generateAvailableInstanceRows(
+	ctx context.Context, sessionExpiration hlc.Timestamp,
+) error {
+	ctx = multitenant.WithTenantCostControlExemption(ctx)
+	// We don't care about the results here.
+	_, _, err := s.reclaimGroup.Do("reclaim-instance-ids", func() (interface{}, error) {
+		err := s.db.Txn(ctx, func(ctx context.Context, txn *kv.Txn) error {
+			// Set the transaction deadline to the session expiration to ensure
+			// transaction commits before the session expires.
+			err := txn.UpdateDeadline(ctx, sessionExpiration)
+			if err != nil {
+				return err
+			}
+
+			// Fetch IDs to claim and delete.
+			toClaim, toDelete, err := s.idsToReclaim(ctx, txn)
```

Comment:
I think we should split the transaction allocating fresh IDs from the transaction that reclaims existing IDs. Reclaiming an existing ID requires checking if the SessionID is alive, and that requires a round trip between the caller and the session's region. That is fine in the claim loop, since the caller is collocated with the session, but it dramatically increases the worst case performance of allocating new IDs.

One idea is we can have two functions `generateAvailableInstanceRows()` and `cleanUpStaleIds()`. `cleanUpStaleIds()` would perform a transaction to read all instance IDs. Outside of the transaction it would call `slstorage.IsAlive()` on the session and would use a conditional put to delete the instance if it is invalid. As long as the background routine calls `cleanUpStaleIds()` before `generateAvailableInstanceRows()`, we can still avoid most unnecessary ID allocations.

Splitting the transaction does force us to pay an extra cross-region read in the background job, but it improves the performance of claiming ids during startup, which is what we need to optimize for.

Reply:
Alternatively, we may be able to use the slstorage.Storage.CachedReader() when checking liveness. That allows us to keep the current structure without needing to pay for cross region reads during startup.

Reply:
> Outside of the transaction it would call slstorage.IsAlive() on the session and would use a conditional put to delete the instance if it is invalid. 

What's the reasoning behind doing those IsAlive() calls outside of the transaction?

> we may be able to use the slstorage.Storage.CachedReader() when checking liveness.

I'm not sure how the cached reader works yet, but if the data in that is stale, wouldn't that cause us to override entries which are actually being used in sql_instances?

Reply:
> What's the reasoning behind doing those IsAlive() calls outside of the transaction?

The motivation is the number of cross region round trips the claim routine waits on. Consider the case where us-central is trying to allocate and europe-west has 3 active sessions. `generateAvailableInstanceRows` would execute one round trip to europe to list the instances in the region, then each IsActive call would block for a round trip checking if the session is still alive. So in total there are 4 round trips in this example, when we only need to pay for 1. This is fine in the background job, but `generateAvailableInstanceRows` is called from the cold start process.

> I'm not sure how the cached reader works yet, but if the data in that is stale, wouldn't that cause us to override entries which are actually being used in sql_instances?

The cached reader is conservative. It returns `true` if it does not know. It only returns `false` if it is certain the session does not exist.

Reply:
I went ahead with the cached reader since that allows us to maintain the current structure.